### PR TITLE
Feature/gbr 41

### PR DESCRIPTION
**Pull request by mikeborozdin**
This is a pull request for https://github.com/caitp/grunt-bower-release/issues/41

It has a number of Git repo files committed inside the `test/e2/` folder. 

Indeed, there are end-to-end tests for new functionality, as well, as a regression test. Unlike previous tests, which are still there, they don't rely upon `grunt test`, instead they can be run using nodeunit. I, for example, can run them using WebStorm.

The way those test work - each of them has a resource folder with a Gruntfile.js. Some of the tests create a git repo on the fly and check whether it contains a newly pushed tag. Other tests has already a git repo with all the necessary tests pre-conditions.

### CODE REVIEW

File: tasks/bower-release.js [hunk 1]
```diff
@@ -162,6 +162,9 @@ module.exports = function(grunt) {
        */
       endpoint.clone(options.endpoint, options.branchName, '.', cloned)
       function cloned(err) {
+        if (err !== null && err !== undefined) {
```
Comment (caitp) [inline tasks/bower-release.js]: `typeof err !== "undefined"`, `_.isUndefined(err)`

File: test/e2e/bower-release-test.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  });
+};
+
+exports['should push a git tag if everything is all right'] = function (test) {
+  test.expect(1);
+
+  process.chdir(projectDir + '/test/e2e/should-push-git-tag/');
+
+  var exec = require('child_process').exec;
+  var async = require('async');
```
Comment (caitp) [inline test/e2e/bower-release-test.js]: async seems to not be in package.json in this CL

**Comment by caitp** [comment]
I'm going to merge your other PR first, so this might become a little bitrotten

**Comment by caitp** [comment]
yeah, it's become bitrotten, needs a rebase

**Comment by mikeborozdin** [comment]
@caitp,

Fixed. `coveralls` reports a decrease in the test coverage. True. It is not able to detect the new code is covered by the end-to-end tests located in `test/e2e/bower-release-test.js`.

There is a possibility of having a unit test by mocking an endpoint and, as a result, increasing the `coveralls` code coverage metric However, this mock will be different from already existing `test-ep`. If we want to go that way, then we'll have to introduce changes to `bower-release`, namely we should allow it to load any arbitrary endpoint via `require`. In this case, we will also solve the current problem of requiring a mock endpoint from the production code.

**Comment by caitp** [comment]
I'm not too worried about losing some test coverage, since the test coverage is awful anyways, and it's not significantly more awful than it already is. This is a big change, it'll take me a bit to go through it, and probably won't happen today

**Comment by mikeborozdin** [comment]
Hi @caitp,

How are you?

I wonder if you have had any time to go over this pull request.

**Comment by caitp** [comment]
I won't be able to go over it today, but I'll try this week

**Comment by caitp** [comment]
and, as for publishing, I'm going to push a new release after this lands

**Comment by mikeborozdin** [comment]
Thank you!

File: tasks/bower-release.js [hunk 3]
```diff
@@ -24,7 +24,7 @@
 
 'use strict';
 
-module.exports = function(grunt) {
+module.exports = function (grunt) {
```
Comment (caitp) [inline tasks/bower-release.js:27]: revert

File: tasks/bower-release.js [hunk 4]
```diff
@@ -68,273 +68,297 @@ module.exports = function(grunt) {
...
-        grunt.log.writeln('  ' + key.cyan + ': ' + options[key].yellow)
-      })
-    }
-
-    /* Read Bower configuration */
-    ['bower.json', 'component.json'].forEach(function (configFile) {
-      if (!bowerJSON && grunt.file.isFile(configFile)) {
-        bowerJSON = grunt.file.readJSON(configFile)
-        bowerFile = configFile
+      if (grunt.option('verbose')) {
```
Comment (caitp) [inline tasks/bower-release.js:71]: don't indent this, it should be 2 spaces from the `grunt.registerMultiTask(...)`

File: tasks/bower-release.js [hunk 5]
```diff
@@ -68,273 +68,297 @@ module.exports = function(grunt) {
...
+
+        /* Make sure stageDir either does not exist, or is a directory */
+        var stat
+        try {
+          stat = fs.statSync(options.stageDir)
+          if (stat && !stat.isDirectory()) {
+            finish(new Error('stageDir is not a directory'))
+            return false
+          }
+        } catch (e) { /* ENOENT */
```
Comment (caitp) [inline tasks/bower-release.js:162]: revert, this was fine before

File: tasks/bower-release.js [hunk 6]
```diff
@@ -68,273 +68,297 @@ module.exports = function(grunt) {
...
          */
-        if (bowerFile) files.push(bowerFile)
-        grunt.file.write(options.stageDir + '/' + bowerFile,
-          JSON.stringify(bowerJSON, null, 2))
-        copyBuildFilesToStage()
-      }
+        endpoint.clone(options.endpoint, options.branchName, '.', cloned)
+        function cloned(err) {
+          if (err !== null && err !== undefined) {
+            failTask(err);
```
Comment (caitp) [inline tasks/bower-release.js:175]: Have you tested this with this kind of behaviour? we don't necessarily care here

File: tasks/bower-release.js [hunk 7]
```diff
@@ -68,273 +68,297 @@ module.exports = function(grunt) {
...
-         * The VCS plugin is not responsible for comprehension of the grunt file
-         * objects, and so we can do that here.
-         */
-        process.chdir(options.stageDir)
-        endpoint.add(files, addedFiles)
-      }
+        function copiedFiles(err, results) {
+          var files = []
+          results.forEach(function (item) {
+            if (item instanceof Array)
```
Comment (caitp) [inline tasks/bower-release.js:254]: we should be using Array.isArray instead here

Comment (caitp) [inline tasks/bower-release.js:254]: also, even though it wasn't there before, should add curly braces around this branch

File: tasks/bower-release.js [hunk 8]
```diff
@@ -68,273 +68,297 @@ module.exports = function(grunt) {
...
         }
-      }
 
-      function tag() {
-        /* Tag name must be valid semver -- but I'm not validating this here. */
-        /* TODO: Validate this here! */
-        var tag = (options.suffixTagWithTimestamp) ? bowerJSON.version + '+' + new Date().getTime() : bowerJSON.version;
-        endpoint.tag(tag, makeTagMsg(options.packageName), function() { tagged(tag) });
-      }
+        function tagged(tag, err, stdout) {
```
Comment (caitp) [inline tasks/bower-release.js:316]: does anything ever pass `stdout` to this?

File: tasks/bower-release.js [hunk 9]
```diff
@@ -68,273 +68,297 @@ module.exports = function(grunt) {
...
-          fs.rmrfSync(options.stageDir)
-        endpoint.tearDown(function(err){
-          if(err)
-            finish(err);
-          finish();
-        });
-      }
+        function failTask(error) {
+          if (!grunt.option('debug')) {
+            process.chdir(startDir);
```
Comment (caitp) [inline tasks/bower-release.js:356]: change to startDir regardless of if in debug mode or not

File: tasks/bower-release.js [hunk 10]
```diff
@@ -68,273 +68,297 @@ module.exports = function(grunt) {
...
-      function makeTagMsg(name) {
-        return name + '@' + bowerJSON.version
+          grunt.fail.fatal(error);
+        }
       }
-    }
- })
+    })
 }
-
```
Comment (caitp) [inline tasks/bower-release.js:340]: revert, a blank line at end of file is needed

File: test/e2e/should-fail-if-tag-exists/repo.git/hooks/pre-push.sample [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+
+remote="$1"
+url="$2"
+
+z40=0000000000000000000000000000000000000000
+
+IFS=' '
+while read local_ref local_sha remote_ref remote_sha
+do
+	if [ "$local_sha" = $z40 ]
```
Comment (caitp) [inline test/e2e/should-fail-if-tag-exists/repo.git/hooks/pre-push.sample:30]: 2-space indentation, even for shell scripts please

**Comment by caitp** [comment]
okay, with comments addressed it looks okay to me

**Comment by caitp** [comment]
hey, if the style nots get addressed I'm happy to merge this, let me know~

**Pull request closed without merging**